context/logger.go

```diff
@@ -0,0 +1,101 @@
+// Copyright 2016 The Linux Foundation
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package context
+
+import (
+	"fmt"
+	"os"
+	"reflect"
+
+	"github.com/Sirupsen/logrus"
+	"golang.org/x/net/context"
+)
+
+var (
+	// G is an alias for GetLogger.
+	//
+	// We may want to define this locally to a package to get package tagged log
+	// messages.
+	G = GetLogger
+
+	// LogEntry provides a public and standard logger instance.
+	LogEntry = logrus.NewEntry(logrus.StandardLogger())
+)
+
+type (
+	loggerKey struct{}
+
+	stderrHook struct{}
```

Comment:
On Thu, Mar 16, 2017 at 04:41:49PM -0700, xiekeyang wrote:
> If it is a popular way for client tools?

POSIX [suggests][1]:

> … standard output (for writing conventional output), and standard error (for writing diagnostic output).

The Linux man pages [have][2]:

> Under normal circumstances every UNIX program has three streams opened for it when it starts up, one for input, one for output, and one for printing diagnostic or error messages… the output stream is referred to as "standard output"; and the error stream is referred to as "standard error".

So I think “normal output to stdout and logging/debug to stderr” is a fairly well-established pattern.  Deciding what is “normal output” and what is “logging/debug” may require more clarity around our intended output, but I'd guess most things which need tunable verbosity are debugging logs which should be going to stderr.

[1]: http://pubs.opengroup.org/onlinepubs/9699919799/functions/stdin.html
[2]: http://man7.org/linux/man-pages/man3/stdout.3.html